packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2277,6 +2277,9 @@ export class KeyringController extends BaseController<
    */
   #updateVault(): Promise<boolean> {
     return this.#withVaultLock(async () => {
+      // Ensure no duplicate accounts are persisted.
+      await this.#assertNoDuplicateAccounts();
```

Comment:
An eventual error thrown here will trigger a rollback in all public functions that mutate the vault. That is, any new account added before this line and after the last vault update will be discarded